backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/shacl/NodeShapeFactory.java

```diff
@@ -0,0 +1,166 @@
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import lombok.NonNull;
+import org.dotwebstack.framework.backend.rdf4j.ValueUtils;
+import org.dotwebstack.framework.backend.rdf4j.shacl.propertypath.PropertyPathFactory;
+import org.eclipse.rdf4j.model.IRI;
+import org.eclipse.rdf4j.model.Literal;
+import org.eclipse.rdf4j.model.Model;
+import org.eclipse.rdf4j.model.Resource;
+import org.eclipse.rdf4j.model.util.Models;
+import org.eclipse.rdf4j.model.vocabulary.RDF;
+import org.eclipse.rdf4j.model.vocabulary.SHACL;
+import org.eclipse.rdf4j.sail.memory.model.MemBNode;
+import org.eclipse.rdf4j.sail.memory.model.MemIRI;
+import org.eclipse.rdf4j.sail.memory.model.MemResource;
+import org.eclipse.rdf4j.sail.memory.model.MemStatement;
+
+public class NodeShapeFactory {
+
+  private NodeShapeFactory() {}
+
+  public static NodeShape createShapeFromModel(@NonNull Model shapeModel, @NonNull IRI identifier) {
+    return NodeShape.builder()
+        .identifier(identifier)
+        .targetClass(ValueUtils.findRequiredPropertyIri(shapeModel, identifier, SHACL.TARGET_CLASS))
+        .propertyShapes(buildPropertyShapes(shapeModel, identifier))
+        .build();
+  }
+
+  private static Map<String, PropertyShape> buildPropertyShapes(Model shapeModel, Resource nodeShape) {
+    /*
+     * The sh:or can occur on multiple levels, either as a direct child of a nodeshape or as a child of
+     * an sh:property here the direct childs of type sh:or on a nodeshape are processed
+     */
+    Map<String, PropertyShape> orShapes = getOrPropertyShapes(shapeModel, nodeShape).stream()
+        .map(shape -> buildPropertyShape(shapeModel, shape))
+        .collect(Collectors.toMap(PropertyShape::getName, Function.identity()));
+
+    Map<String, PropertyShape> propertyShapes = Models.getPropertyResources(shapeModel, nodeShape, SHACL.PROPERTY)
+        .stream()
+        .map(shape -> {
+          if (ValueUtils.isPropertyPresent(shapeModel, shape, SHACL.OR)) {
+            /* This processes sh:or constructs that are childs of the sh:property */
+            return getOrPropertyShapes(shapeModel, shape);
+          } else {
+            return Collections.singletonList(shape);
+          }
+        })
+        .flatMap(List::stream)
+        .map(shape -> buildPropertyShape(shapeModel, shape))
+        .collect(Collectors.toMap(PropertyShape::getName, Function.identity()));
+
+    return Stream.concat(orShapes.entrySet()
+        .stream(),
+        propertyShapes.entrySet()
+            .stream())
+        .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+  }
+
+  private static List<Resource> getOrPropertyShapes(Model shapeModel, Resource nodeShape) {
+    return Models.getPropertyResources(shapeModel, nodeShape, SHACL.OR)
+        .stream()
+        .map(or -> unwrapOrStatements(shapeModel, or).stream()
+            .peek(resource ->
+            /*
```

Comment:
Typo en zou iets korter kunnen, zoiets?
```suggestion
            /*
             * All the individual childs under an sh:or statement are enriched with the shared values on the same
             * level as the sh:or. The sh:or is excluded from this enrichment. These enrichments are
             * added to the model for later reuse.
             */
```